nemo/collections/nlp/modules/common/megatron/transformer.py

```diff
@@ -1718,79 +2004,64 @@ def forward(
-                    past = layer_past[index]
-                hidden_states = layer(
+            rng_context = nullcontext()
+
+        with rng_context:
+            if self.activations_checkpoint_granularity == 'full':
+                hidden_states = self._checkpointed_forward(
                     hidden_states,
                     attention_mask,
-                    encoder_output=encoder_output,
-                    enc_dec_attn_mask=enc_dec_attn_mask,
-                    layer_past=past,
-                    get_key_value=get_key_value,
-                    set_inference_key_value_memory=set_inference_key_value_memory,
-                    inference_max_sequence_len=inference_max_sequence_len,
-                    rotary_pos_emb=rotary_pos_emb,
-                    position_bias=position_bias,
-                    encoder_decoder_position_bias=encoder_decoder_position_bias,
+                    encoder_output,
+                    enc_dec_attn_mask,
+                    rotary_pos_emb,
+                    position_bias,
+                    encoder_decoder_position_bias,
                 )
-                if get_key_value:
-                    hidden_states, present = hidden_states
-                    presents.append(present)
-                if self.position_embedding_type == 'relative':
+
+                if type(hidden_states) is tuple:
                     if len(hidden_states) == 2:
                         hidden_states, position_bias = hidden_states
                     elif len(hidden_states) == 3:
                         hidden_states, position_bias, encoder_decoder_position_bias = hidden_states
                     else:
                         raise IndexError('Hidden_states needs to be tuple containing 2 or 3 elements.')
+            else:
+                if get_key_value:
+                    presents = []
+                for index in range(self.num_layers):
+                    layer = self._get_layer(index)
+                    past = None
+                    if layer_past is not None:
+                        past = layer_past[index]
+                    hidden_states = layer(
+                        hidden_states,
+                        attention_mask,
+                        encoder_output=encoder_output,
+                        enc_dec_attn_mask=enc_dec_attn_mask,
+                        layer_past=past,
+                        get_key_value=get_key_value,
+                        set_inference_key_value_memory=set_inference_key_value_memory,
+                        inference_max_sequence_len=inference_max_sequence_len,
+                        rotary_pos_emb=rotary_pos_emb,
+                        position_bias=position_bias,
+                        encoder_decoder_position_bias=encoder_decoder_position_bias,
+                    )
 
         # Final layer norm.
         if self.post_process:
-            # Reverting data format change [s b h] --> [b s h].
```

Comment:
if we assume the order [s b h], RETRO encoder/decoder needs to be fixed as it assuems [b s h]